README
#echievements on FreeNode IRC


ChangeLog
Dec 22 2012  Mike Blumenkrantz

	* use border hook for new window achievements so we don't get multiples
	* fix module_init_end handler to detect if echievements was loaded late
	* switch config version macros to match new version scheme

Dec 21 2012  Igor Murzov

	* Updated README

Dec 21 2012  Mike Blumenkrantz

	* Updated COPYING; Echievements is BSD-2
	* v2 release


src/e_mod_main.h
#ifndef E_MOD_MAIN_H
#define E_MOD_MAIN_H

#ifdef HAVE_CONFIG_H
# include "config.h"
#endif

#include <e.h>
#include <Etrophy.h>
#include <E_Notify.h>

#ifdef ENABLE_NLS
# include <libintl.h>
# define D_(str) dgettext(PACKAGE, str)
# define DP_(str, str_p, n) dngettext(PACKAGE, str, str_p, n)
#else
# define bindtextdomain(domain,dir)
# define bind_textdomain_codeset(domain,codeset)
# define D_(str) (str)
# define DP_(str, str_p, n) (str_p)
#endif

#define N_(str) (str)

#define MOD_CONFIG_FILE_EPOCH 0x0000
#define MOD_CONFIG_FILE_GENERATION 0x01
#define MOD_CONFIG_FILE_VERSION \
   ((MOD_CONFIG_FILE_EPOCH << 16) | MOD_CONFIG_FILE_GENERATION)

#define XSTR(X) #X
#define STRINGIFY(X) XSTR(X)

// need to update these whenever default profile updates I guess
#define DEFAULT_BINDING_COUNT_ACPI 11
#define DEFAULT_BINDING_COUNT_KEY 65
#define DEFAULT_BINDING_COUNT_MOUSE 6
#define DEFAULT_BINDING_COUNT_SIGNAL 28

#define ECH(X) ECHIEVEMENT_ID_##X
/* See instructions in echievements.x for adding new ones */
typedef enum
{
   ECH(SHELF_POSITIONS),
   ECH(NOTHING_ELSE_MATTERS),
   ECH(FEAR_OF_THE_DARK),
   ECH(CAVE_DWELLER),
   ECH(WINDOW_ENTHUSIAST),
   ECH(WINDOW_LOVER),
   ECH(WINDOW_STALKER),
   ECH(WINDOW_MOVER),
   ECH(WINDOW_OCD),
   ECH(MOUSE_RUNNER),
   ECH(MOUSE_MARATHONER),
   ECH(MOUSE_HERO),
   ECH(MIGHTY_MOUSE),
   ECH(WINDOW_HAULER),
   ECH(WINDOW_SLINGER),
   ECH(WINDOW_SHERPA),
   ECH(PHYSICIST),
   ECH(QUICKDRAW),
   ECH(OPAQUE),
   ECH(KEYBOARD_USER),
   ECH(KEYBOARD_TAPPER),
   ECH(KEYBOARD_ABUSER),
   ECH(KEYBOARD_NINJA),
   ECH(CLICKER),
   ECH(SUPER_CLICKER),
   ECH(SUPER_DUPER_CLICKER),
   ECH(CLICK_MANIAC),
   ECH(SLEEPER),
   ECH(HALTER),
   ECH(EDGY),
   ECH(EDGIER),
   ECH(EDGIEST),
   ECH(EDGAR),
   ECH(WHEELY),
   ECH(ROLLIN),
   ECH(ROLY_POLY),
   ECH(SIGNALLER),
   ECH(SIGNAL_CONTROLLER),
   ECH(SIGNAL_MASTER),
   ECH(SIGURD),
   ECH(NOT_SO_INCOGNITO),
   ECH(DUALIST),
   ECH(THRICE),
   ECH(GOING_HD),
   ECH(REAL_ESTATE_MOGUL),
   ECH(MAXIMUM_DEFINITION),
   ECH(TERMINOLOGIST),
   ECH(LIFE_ON_THE_EDGE), //FIXME: this will be granted immediately until module goes into core :/
   ECH(BILINGUAL),
   ECH(POLYGLOT),
   ECH(GADGETEER),
   ECH(INSPECTOR_GADGET),
   ECH(GADGET_HACKWRENCH),
   ECH(SECURITY_CONSCIOUS),
   ECH(SECURITY_NUT),
   ECH(CHIEF_OF_SECURITY),
   ECH(TILED),
   ECH(PERSISTENT), //FIXME: timer does not persist across restarts :/
   ECH(NEVER_GONNA_GIVE_YOU_UP),
   ECH(LAST)
} Echievement_Id;

typedef struct Echievement
{
   EINA_INLIST;
   Echievement_Id id;
   Etrophy_Trophy *trophy;
   Eina_List *handlers;
   Ecore_Cb mouse_hook;
   struct
   {
      Evas_Object *icon;
      Evas_Object *label;
   } dialog;
   void *data; // echievement data
   Eina_Bool bh_handlers : 1; //handlers are border hooks
} Echievement;

typedef void (*Echievement_Init_Cb)(Echievement*);

typedef struct Mod
{
   E_Config_Dialog *cfd;
   Evas_Object *cfd_list[2]; //list of earned trophies, list of all trophies
   Evas_Object *label; //label of total points
   Evas_Object_Event_Cb obj_del_cb; //for deleting list object callbacks

   E_Module *module;
   Eina_Hash *trophies; // for fast lookups
   Echievement *trophies_list; // head for iterating
   Eina_Inlist *itr; // current iterator position if needed
   struct
   {
      int x, y;
      int dx, dy;
      Eina_List *hooks;
   } mouse;
   struct
   {
      Eina_List *timers;
   } desklock;
   Eina_Bool module_init_end : 1;
} Mod;

typedef struct Config
{
   unsigned int config_version;
   Etrophy_Gamescore *gs;
} Config;

extern Mod *mod;
extern Config *ech_config;

E_Config_Dialog *e_int_config_echievements(E_Container *con, const char *params);
EINTERN void ech_cfg_ech_update(Echievement *ec);
EINTERN void ech_cfg_ech_add(Echievement *ec);

void ech_init(void);
void ech_shutdown(void);

unsigned int ech_bindings_check_keys(void);
unsigned int ech_bindings_check_mouse(void);
unsigned int ech_bindings_check_wheel(void);
unsigned int ech_bindings_check_edge(void);
unsigned int ech_bindings_check_acpi(void);
unsigned int ech_bindings_check_signal(void);

EINTERN Eina_List *ech_language_enumerate(void);
#endif


src/e_mod_config.c
#include "e_mod_main.h"


static void
_free_data(E_Config_Dialog *cfd EINA_UNUSED, E_Config_Dialog_Data *cfdata EINA_UNUSED)
{
   mod->cfd = NULL;
   mod->label = mod->cfd_list[0] = mod->cfd_list[1] = NULL;
}
/*
static void *
_create_data(E_Config_Dialog *cfd EINA_UNUSED)
{
   return NULL;
}

static int
_basic_apply(E_Config_Dialog *cfd EINA_UNUSED, E_Config_Dialog_Data *cfdata EINA_UNUSED)
{
   return 0;
}

static int
_basic_check_changed(E_Config_Dialog *cfd EINA_UNUSED, E_Config_Dialog_Data *cfdata EINA_UNUSED)
{
   return 0;
}
*/

static void
_obj_del(Echievement *ec, Evas *e EINA_UNUSED, Evas_Object *obj, void *event_info EINA_UNUSED)
{
   if (obj == ec->dialog.icon)
     ec->dialog.icon = NULL;
   else
     ec->dialog.label = NULL;
}

static Evas_Object *
_item_create(Evas *evas, Echievement *ec, Eina_Bool save)
{
   Evas_Object *table, *icon, *label;
   char progress[128];
   unsigned int goal, counter;

   etrophy_trophy_goal_get(ec->trophy, &goal, &counter);
   table = e_widget_frametable_add(evas, D_(etrophy_trophy_name_get(ec->trophy)), 1);

   icon = e_widget_image_add_from_file(evas, PACKAGE_DATA_DIR "/trophy.png",
                                       64, 64);
   if (save) ec->dialog.icon = icon;
   evas_object_event_callback_add(icon, EVAS_CALLBACK_DEL, (Evas_Object_Event_Cb)_obj_del, ec);
   e_widget_frametable_object_append(table, icon, 0, 0, 1, 2, 1, 1, 1, 0);
   label = e_widget_label_add(evas, D_(etrophy_trophy_description_get(ec->trophy)));
   e_widget_frametable_object_append(table, label, 1, 0, 3, 1, 1, 1, 1, 0);

   if (etrophy_trophy_earned_get(ec->trophy))
     {
        char date[256];
        char buf[512];
        time_t t0;

        t0 = etrophy_trophy_date_get(ec->trophy);
        strftime(date, sizeof(date), "%d %B %Y - %r", localtime(&t0));
        snprintf(buf, sizeof(buf), D_("Achieved: %s"), date);
        label = e_widget_label_add(evas, buf);
     }
   else if (goal < 2)
     label = e_widget_label_add(evas, D_("Not achieved"));
   else
     {
        snprintf(progress, sizeof(progress), D_("Progress: %u/%u"), counter, goal);
        label = e_widget_label_add(evas, progress);
     }
   if (save) ec->dialog.label = label;
   evas_object_event_callback_add(label, EVAS_CALLBACK_DEL, (Evas_Object_Event_Cb)_obj_del, ec);
   e_widget_frametable_object_append(table, label, 1, 1, 3, 1, 1, 1, 1, 0);

   return table;
}

static Evas_Object *
_basic_create(E_Config_Dialog *cfd EINA_UNUSED, Evas *evas, E_Config_Dialog_Data *cfdata EINA_UNUSED)
{
   Evas_Object *toolbook, *list, *item, *sf;
   Echievement *ec;
   int mw, mh, mww;

   toolbook = e_widget_toolbook_add(evas, 48 * e_scale, 48 * e_scale);

   /* FIXME it should be scrollable and all initial items should
      be visible */
   mod->cfd_list[0] = list = e_widget_list_add(evas, 1, 0);
   EINA_INLIST_FOREACH(EINA_INLIST_GET(mod->trophies_list), ec)
     {
        if (!etrophy_trophy_earned_get(ec->trophy)) continue;
        item = _item_create(evas, ec, EINA_FALSE);
        e_widget_list_object_append(list, item, 1, 1, 0.5);
     }
   e_widget_size_min_get(list, &mw, &mh);

   if (mw < 320) mw = 320;
   if (mh < 220) mh = 220;

   evas_object_resize(list, mw, mh);
   sf = e_widget_scrollframe_simple_add(evas, list);
   e_widget_size_min_set(sf, 320, 220);
   e_widget_toolbook_page_append(toolbook, NULL, D_("My Trophies"),
                                 sf, 1, 1, 1, 1, 0.5, 0.0);

   mod->cfd_list[1] = list = e_widget_list_add(evas, 1, 0);

   EINA_INLIST_FOREACH(EINA_INLIST_GET(mod->trophies_list), ec)
     {
        if ((!etrophy_trophy_earned_get(ec->trophy)) && (!etrophy_trophy_visibility_get(ec->trophy))) continue;
        item = _item_create(evas, ec, EINA_TRUE);
        e_widget_list_object_append(list, item, 1, 1, 0.5);
     }
   e_widget_size_min_get(list, &mww, &mh);

   if (mww < mw) mww = mw;
   if (mh < 220) mh = 220;

   evas_object_resize(list, mww, mh);
   sf = e_widget_scrollframe_simple_add(evas, list);
   e_widget_size_min_set(sf, 320, 220);
   e_widget_toolbook_page_append(toolbook, NULL, D_("All Trophies"),
                                 sf, 1, 1, 1, 1, 0.5, 0.0);

   e_widget_toolbook_page_show(toolbook, 0);
   list = e_widget_list_add(evas, 0, 0);
   e_widget_list_object_append(list, toolbook, 1, 1, 0.5);
   {
      char buf[1024];
      unsigned score = etrophy_gamescore_trophies_points_get(ech_config->gs);

      /* FIXME: this can look better */
      snprintf(buf, sizeof(buf), DP_("%u Echievement Point", "%u Echievement Points", score), score);
      mod->label = item = e_widget_label_add(evas, buf);
      e_widget_list_object_append(list, item, 0, 0, 0.5);
   }
   return list;
}

static void
_totals_update(void)
{
   char buf[1024];
   unsigned score = etrophy_gamescore_trophies_points_get(ech_config->gs);

   snprintf(buf, sizeof(buf), DP_("%u Echievement Point", "%u Echievement Points", score), score);
   e_widget_label_text_set(mod->label, buf);
}

E_Config_Dialog *
e_int_config_echievements(E_Container *con, const char *params EINA_UNUSED)
{
   E_Config_Dialog *cfd = NULL;
   E_Config_Dialog_View *v = NULL;
   char buf[4096];

   if (e_config_dialog_find("Echievements", "extensions/echievements"))
     return NULL;

   mod->obj_del_cb = (Evas_Object_Event_Cb)_obj_del;
   v = E_NEW(E_Config_Dialog_View, 1);
   v->free_cfdata = _free_data;
   v->basic.create_widgets = _basic_create;
/*
   v->create_cfdata = _create_data;
   v->basic.apply_cfdata = _basic_apply;
   v->basic.check_changed = _basic_check_changed;
*/
   snprintf(buf, sizeof(buf), "%s/e-module-echievements.edj",
            mod->module->dir);

   cfd = e_config_dialog_new(con, D_("Echievements"), "Echievements",
                             "extensions/echievements", buf, 0, v, NULL);

   e_dialog_resizable_set(cfd->dia, 1);
   mod->cfd = cfd;
   return cfd;
}

EINTERN void
ech_cfg_ech_update(Echievement *ec)
{
   char progress[128];
   unsigned int goal, counter;

   if (!mod->cfd) return;
   if ((!ec->dialog.icon) || (!ec->dialog.label)) return;
   etrophy_trophy_goal_get(ec->trophy, &goal, &counter);
   if (etrophy_trophy_earned_get(ec->trophy))
     {
        char date[256];
        char buf[512];
        time_t t0;

        t0 = etrophy_trophy_date_get(ec->trophy);
        strftime(date, sizeof(date), "%d %B %Y - %r", localtime(&t0));
        snprintf(buf, sizeof(buf), D_("Achieved: %s"), date);
        e_widget_label_text_set(ec->dialog.label, buf);
     }
   else if (goal < 2)
     e_widget_label_text_set(ec->dialog.label, D_("Not achieved"));
   else
     {
        snprintf(progress, sizeof(progress), D_("Progress: %u/%u"), counter, goal);
        e_widget_label_text_set(ec->dialog.label, progress);
     }
   e_widget_frametable_object_repack(e_widget_parent_get(ec->dialog.label), ec->dialog.label, 1, 1, 3, 1, 1, 1, 1, 0);
   _totals_update();
}

EINTERN void
ech_cfg_ech_add(Echievement *ec)
{
   Evas *e;
   Evas_Object *item;
   int mh;
   int w;

   if (!mod->cfd) return;
   /* use "all trophies" list current width
    * since it's guaranteed to be correct for the moment
    */
   evas_object_geometry_get(mod->cfd_list[1], NULL, NULL, &w, NULL);
   e = evas_object_evas_get(mod->cfd_list[0]);
   item = _item_create(e, ec, EINA_FALSE);
   e_widget_list_object_append(mod->cfd_list[0], item, 1, 1, 0.5);
   e_widget_size_min_get(mod->cfd_list[0], NULL, &mh);

   evas_object_resize(mod->cfd_list[0], w, mh);

   if (ec->dialog.icon || ec->dialog.label) return;

   item = _item_create(e, ec, EINA_TRUE);
   e_widget_list_object_append(mod->cfd_list[1], item, 1, 1, 0.5);
   e_widget_size_min_get(mod->cfd_list[1], NULL, &mh);
   evas_object_resize(mod->cfd_list[1], w, mh);
   _totals_update();
}


src/e_mod_main.c
#include "e_mod_main.h"

EAPI E_Module_Api e_modapi = {E_MODULE_API_VERSION, "Echievements"};
static E_Config_DD *conf_edd = NULL;
static Eina_List *handlers = NULL;

EINTERN Mod *mod = NULL;
EINTERN Config *ech_config = NULL;

static void
_e_mod_ech_config_free(void)
{
   etrophy_gamescore_free(ech_config->gs);
   E_FREE(ech_config);
}

static void
_e_mod_ech_config_load(void)
{
#undef T
#undef D
   conf_edd = E_CONFIG_DD_NEW("Config", Config);
   #define T Config
   #define D conf_edd
   E_CONFIG_VAL(D, T, config_version, UINT);
   E_CONFIG_SUB(D, T, gs, etrophy_gamescore_edd_get());

   ech_config = e_config_domain_load("module.echievements", conf_edd);
   if (ech_config)
     {
        if (!e_util_module_config_check("Echievements", ech_config->config_version, MOD_CONFIG_FILE_VERSION))
          _e_mod_ech_config_free();
     }

   if (!ech_config)
     {
        ech_config = E_NEW(Config, 1);
        ech_config->config_version = (MOD_CONFIG_FILE_EPOCH << 16);
     }
   if (!ech_config->gs)
     ech_config->gs = etrophy_gamescore_new("echievements");
}

static Eina_Bool
_e_mod_ech_mod_init_end_cb(void *d EINA_UNUSED, int type EINA_UNUSED, void *ev EINA_UNUSED)
{
   ecore_event_handler_del(eina_list_data_get(handlers));
   handlers = eina_list_remove_list(handlers, handlers);
   mod->module_init_end = 1;
   return ECORE_CALLBACK_RENEW;
}

static Eina_Bool
_e_mod_ech_mouse_move_cb(void *d EINA_UNUSED, int type EINA_UNUSED, Ecore_Event_Mouse_Move *ev)
{
   if ((mod->mouse.x != mod->mouse.y) || (mod->mouse.x != -1))
     {
        Echievement *ec;
        Eina_List *l;

        mod->mouse.dx = ev->root.x - mod->mouse.x, mod->mouse.dy = ev->root.y - mod->mouse.y;
        EINA_LIST_FOREACH(mod->mouse.hooks, l, ec)
          if (ec->mouse_hook) ec->mouse_hook(ec);
     }
   mod->mouse.x = ev->root.x, mod->mouse.y = ev->root.y;
   return ECORE_CALLBACK_RENEW;
}

static Eina_Bool
_e_mod_ech_desklock_timer_cb(void *d EINA_UNUSED)
{
   mod->desklock.timers = eina_list_remove_list(mod->desklock.timers, mod->desklock.timers);
   return EINA_FALSE;
}

static Eina_Bool
_e_mod_ech_desklock_cb(void *d EINA_UNUSED, int type EINA_UNUSED, E_Event_Desklock *ev)
{
   if (!ev->on)
     mod->desklock.timers = eina_list_append(mod->desklock.timers,
       ecore_timer_add(60 * 60, _e_mod_ech_desklock_timer_cb, NULL));

   return ECORE_CALLBACK_RENEW;
}

EAPI void *
e_modapi_init(E_Module *m)
{
   char buf[4096];

   bindtextdomain(PACKAGE, LOCALEDIR);
   bind_textdomain_codeset(PACKAGE, "UTF-8");

   snprintf(buf, sizeof(buf), "%s/e-module-echievements.edj", m->dir);

   e_configure_registry_category_add("extensions", 80, D_("Extensions"),
                                     NULL, "preferences-extensions");
   e_configure_registry_item_add("extensions/echievements", 110, D_("Echievements"),
                                 NULL, buf, e_int_config_echievements);

   etrophy_init();
   e_notification_init();

   mod = E_NEW(Mod, 1);
   mod->module = m;
   mod->mouse.x = mod->mouse.y = -1;

   E_LIST_HANDLER_APPEND(handlers, E_EVENT_MODULE_INIT_END, _e_mod_ech_mod_init_end_cb, NULL);
   E_LIST_HANDLER_APPEND(handlers, ECORE_EVENT_MOUSE_MOVE, _e_mod_ech_mouse_move_cb, NULL);
   E_LIST_HANDLER_APPEND(handlers, E_EVENT_DESKLOCK, _e_mod_ech_desklock_cb, NULL);

   _e_mod_ech_config_load();
   ech_init();

   return m;
}

EAPI int
e_modapi_shutdown(E_Module *m EINA_UNUSED)
{
   e_configure_registry_item_del("extensions/echievements");

   e_configure_registry_category_del("extensions");

   E_FN_DEL(e_object_del, mod->cfd);
   ech_shutdown();
   e_config_domain_save("module.echievements", conf_edd, ech_config);
   _e_mod_ech_config_free();
   E_CONFIG_DD_FREE(conf_edd);
   eina_list_free(mod->mouse.hooks);
   E_FREE_LIST(mod->desklock.timers, ecore_timer_del);
   E_FREE(mod);
   E_FREE_LIST(handlers, ecore_event_handler_del);
   e_notification_shutdown();
   etrophy_shutdown();
   return 1;
}

EAPI int
e_modapi_save(E_Module *m EINA_UNUSED)
{
   e_config_domain_save("module.echievements", conf_edd, ech_config);
   return 1;
}

